enhancements/machine-api/out-of-tree-provider-support.md

```diff
@@ -0,0 +1,232 @@
+	// and will concider the value of the infrastructure to be "external" - essentially
+	// disabling any further evaluation.
+	//
+	// This will allow to delegate task of processing platform specific bits
+	// to another component.
+	// +optional
+  External bool `json:"external"`
+  ...
+```
+
+To ensure a smooth transition from the `in-tree` selection, `library-go` [implementation](https://github.com/openshift/library-go/blob/master/pkg/operator/configobserver/cloudprovider/observe_cloudprovider.go) will include an additional argument for the `cloudProviderObserver` structure, which will allow to preserve initial configuration strategy, until the new `observeExternal` flag will be set to `false`:
+
+```go
+type cloudProviderObserver struct {
+	targetNamespaceName     string
+	cloudProviderNamePath   []string
+	cloudProviderConfigPath []string
+	observeExternal         bool
+}
+```
+
+The `observeExternal` field will determine, if `status.platformStatus.external` field from the `Infrastructure` resource will be taken in account in establishing needed flags for a Kubernetes binary, such as some `cloud-controller-manager`, `kube-controller-manager`, etc. It will make the observer type to ignore content of the `cloud-config` file, and force the provider to be set as `external` if the option is not selected.
+
+#### Kubelet
+
+Openshift manages `kubelet` configuration with `machine-config-operator`. The actual `cloud-config` arguments are passed externally from the `Infrastructure` resource on https://github.com/openshift/machine-config-operator/blob/2a51e49b0835bbd3ac60baa685299e86777b064f/pkg/controller/template/render.go#L310-L357, where the support for `external` cloud provider will be added. For each platform with `external` configuration, the `cloud-config` flag will not be set.
+
+#### Kube-controller-manager
+
+This configuration is managed by the `cluster-kube-controller-manager-operator` code and will need to integrate the `library-go` changes, described above in the operator [configuration](https://github.com/openshift/cluster-kube-controller-manager-operator/blob/990ea3f2ace13be6579c00a889842c3f5d3e756a/pkg/operator/configobservation/configobservercontroller/observe_config_controller.go#L82-L85).
+
+#### Kube-api-server
+
+No changes for the current configuration are needed
+
+### Operator resource management strategy
+
+Every cloud manages it's configuration differently, but share a common overall structure.
+
+1. `cloud-controller-manager` and related resources are deployed under one namespace.
+2. Each pod (`ccm`, `cloud-node-manager`, etc.) supports deployment with a `DaemonSet`.
+3. The pods are expected to run on `Nodes` marked as a control-plane with annotation:
+```yaml
+    node-role.kubernetes.io/master: ""
+```
+4. The pods are sharing common command line arguments with the `kcm` pods, such as `--cloud-provider` and `--cloud-config`.
+
+Proposed operator will not diverge from the upstream resource structuring:
```

Comment:
What does this mean?

Reply:
Sorry, I only meant to say the operator will deploy the same set of workloads decided to be used in the upstream repositories. This way our fork for \<some\>cloud-provider will not diverge in the structuring from the upstream until it is decided otherwise.

If they are deploying additional service for some newly introduced feature or set a new port for metrics, we'd like to follow their implementation and avoid possible collisions with our own components, previously structured differently

Reply:
This sounds ambitious. Expect that our manifests we seriously diverge from anything upstream provides. We might follow loosely their structure, but won't be able to use their manifests directly.